File: onMyTssAndroid/app/src/main/java/com/example/onmytss/domain/calculator/BodyBatteryCalculator.kt
package com.example.onmytss.domain.calculator

import com.example.onmytss.domain.Constants
import com.example.onmytss.domain.model.enums.ReadinessLevel
import com.example.onmytss.domain.model.enums.Trend

object BodyBatteryCalculator {

    fun calculateScore(tsb: Double): Int {
        val score: Double = when {
            tsb <= Constants.TSB_FOR_MIN_SCORE -> 0.0
            tsb >= Constants.TSB_FOR_MAX_SCORE -> 100.0
            tsb < 0 -> {
                // Linear from -35 to 0 maps to 0 to 50
                ((tsb - Constants.TSB_FOR_MIN_SCORE) / (-Constants.TSB_FOR_MIN_SCORE)) * 50.0
            }
            else -> {
                // Linear from 0 to 25 maps to 50 to 100
                50.0 + (tsb / Constants.TSB_FOR_MAX_SCORE) * 50.0
            }
        }
        return score.toInt().coerceIn(0, 100)
    }

    fun calculateScoreWithModifiers(tsb: Double, hrvModifier: Double? = null, rhrModifier: Double? = null): Int {
        val baseScore = calculateScore(tsb).toDouble()
        val combinedModifier = PhysiologyModifier.calculateCombinedModifier(hrvModifier, rhrModifier)
        val modifiedScore = (baseScore + combinedModifier).coerceIn(0.0, 100.0)
        return modifiedScore.toInt()
    }

    fun getReadinessLevel(score: Int): ReadinessLevel {
        return when (score) {
            in 0 until 20 -> ReadinessLevel.VERY_LOW
            in 20 until 40 -> ReadinessLevel.LOW
            in 40 until 60 -> ReadinessLevel.MEDIUM
            in 60 until 80 -> ReadinessLevel.GOOD
            else -> ReadinessLevel.EXCELLENT
        }
    }

    fun getScoreChangeDescription(previousScore: Int, currentScore: Int): String {
        val change = currentScore - previousScore
        return when {
            change > 10 -> "↑ Significantly improved (+$change)"
            change > 3 -> "↑ Improved (+$change)"
            change > 0 -> "→ Slightly improved (+$change)"
            change == 0 -> "→ No change"
            change > -3 -> "→ Slightly decreased ($change)"
            change > -10 -> "↓ Decreased ($change)"
            else -> "↓ Significantly decreased ($change)"
        }
    }

    fun calculateTrend(scores: List<Int>): Trend {
        if (scores.size < 3) return Trend.STABLE

        val n = scores.size.toDouble()
        val x = scores.indices.map { it.toDouble() }
        val y = scores.map { it.toDouble() }

        val sumX = x.sum()
        val sumY = y.sum()
        val sumXY = x.zip(y).sumOf { it.first * it.second }
        val sumX2 = x.sumOf { it * it }

        val denominator = n * sumX2 - sumX * sumX
        if (denominator == 0.0) return Trend.STABLE

        val slope = (n * sumXY - sumX * sumY) / denominator

        return when {
            slope > 2.0 -> Trend.IMPROVING_FAST
            slope > 0.5 -> Trend.IMPROVING
            slope < -2.0 -> Trend.DECLINING_FAST
            slope < -0.5 -> Trend.DECLINING
            else -> Trend.STABLE
        }
    }
}

File: onMyTss/onMyTss/Services/BodyBatteryCalculator.swift
//
//  BodyBatteryCalculator.swift
//  onMyTss
//
//  Created by Claude Code
//

import Foundation

/// Calculator for Body Battery score (0-100 scale)
/// Converts Training Stress Balance (TSB) to an intuitive readiness score
struct BodyBatteryCalculator {

    // MARK: - Score Calculation

    /// Convert TSB to Body Battery score (0-100)
    /// TSB typically ranges from -50 (very fatigued) to +30 (very fresh)
    /// We map this to 0-100 scale with 50 as neutral (TSB = 0)
    ///
    /// Mapping:
    /// TSB = -35 or lower → Score = 0 (completely depleted)
    /// TSB = 0 → Score = 50 (neutral)
    /// TSB = +25 or higher → Score = 100 (fully charged)
    static func calculateScore(from tsb: Double) -> Int {
        // Linear interpolation between TSB ranges
        let score: Double

        if tsb <= Constants.tsbForMinScore {
            score = 0
        } else if tsb >= Constants.tsbForMaxScore {
            score = 100
        } else if tsb < 0 {
            // Piecewise-linear: TSB in [minTSB, 0] maps to [0, 50], pinning neutral TSB=0 at 50
            score = ((tsb - Constants.tsbForMinScore) / -Constants.tsbForMinScore) * 50
        } else {
            // Piecewise-linear: TSB in [0, maxTSB] maps to [50, 100]
            score = 50 + (tsb / Constants.tsbForMaxScore) * 50
        }

        // Clamp to 0-100 and round to integer
        return Int(score.clamped(to: 0...100))
    }

    /// Calculate score with HRV/RHR modifiers applied
    /// Combines base TSB score with physiological recovery signals
    static func calculateScoreWithModifiers(
        tsb: Double,
        hrvModifier: Double? = nil,
        rhrModifier: Double? = nil
    ) -> Int {
        let baseScore = Double(calculateScore(from: tsb))

        // Calculate combined modifier from HRV and RHR
        let combinedModifier = PhysiologyModifier.calculateCombinedModifier(
            hrvModifier: hrvModifier,
            rhrModifier: rhrModifier
        )

        // Apply modifier to base score and clamp to 0-100
        let modifiedScore = (baseScore + combinedModifier).clamped(to: 0...100)

        return Int(modifiedScore)
    }

    // MARK: - Score Interpretation

    /// Get readiness level based on score
    static func getReadinessLevel(score: Int) -> ReadinessLevel {
        switch score {
        case 0..<20:
            return .veryLow
        case 20..<40:
            return .low
        case 40..<60:
            return .medium
        case 60..<80:
            return .good
        default:
            return .excellent
        }
    }

    /// Get score change description
    static func getScoreChangeDescription(previousScore: Int, currentScore: Int) -> String {
        let change = currentScore - previousScore

        if change > 10 {
            return "↑ Significantly improved (+\(change))"
        } else if change > 3 {
            return "↑ Improved (+\(change))"
        } else if change > 0 {
            return "→ Slightly improved (+\(change))"
        } else if change == 0 {
            return "→ No change"
        } else if change > -3 {
            return "→ Slightly decreased (\(change))"
        } else if change > -10 {
            return "↓ Decreased (\(change))"
        } else {
            return "↓ Significantly decreased (\(change))"
        }
    }

    // MARK: - Trend Analysis

    /// Calculate trend over last N days
    static func calculateTrend(scores: [Int]) -> Trend {
        guard scores.count >= 3 else { return .stable }

        // Calculate simple linear regression slope
        let n = Double(scores.count)
        let x = Array(0..<scores.count).map { Double($0) }
        let y = scores.map { Double($0) }

        let sumX = x.reduce(0, +)
        let sumY = y.reduce(0, +)
        let sumXY = zip(x, y).map(*).reduce(0, +)
        let sumX2 = x.map { $0 * $0 }.reduce(0, +)

        let denominator = n * sumX2 - sumX * sumX
        guard denominator != 0 else { return .stable }
        let slope = (n * sumXY - sumX * sumY) / denominator

        // Interpret slope
        if slope > 2.0 {
            return .improvingFast
        } else if slope > 0.5 {
            return .improving
        } else if slope < -2.0 {
            return .decliningFast
        } else if slope < -0.5 {
            return .declining
        } else {
            return .stable
        }
    }
}

// MARK: - Readiness Level

enum ReadinessLevel: String {
    case veryLow = "Very Low"
    case low = "Low"
    case medium = "Medium"
    case good = "Good"
    case excellent = "Excellent"

    var description: String {
        switch self {
        case .veryLow:
            return "Severely fatigued. Rest is critical."
        case .low:
            return "Fatigued. Focus on recovery."
        case .medium:
            return "Moderate readiness. Light training okay."
        case .good:
            return "Good readiness for training."
        case .excellent:
            return "Excellent! Ready for hard efforts."
        }
    }

    var emoji: String {
        switch self {
        case .veryLow:
            return "😴"
        case .low:
            return "😓"
        case .medium:
            return "😐"
        case .good:
            return "🙂"
        case .excellent:
            return "💪"
        }
    }
}

// MARK: - Trend

enum Trend: String {
    case improvingFast = "Rapidly Improving"
    case improving = "Improving"
    case stable = "Stable"
    case declining = "Declining"
    case decliningFast = "Rapidly Declining"

    var description: String {
        switch self {
        case .improvingFast:
            return "Your readiness is rapidly improving."
        case .improving:
            return "Your readiness is steadily improving."
        case .stable:
            return "Your readiness is stable."
        case .declining:
            return "Your readiness is declining."
        case .decliningFast:
            return "Your readiness is declining rapidly. Consider more recovery."
        }
    }

    var arrow: String {
        switch self {
        case .improvingFast:
            return "⬆️⬆️"
        case .improving:
            return "⬆️"
        case .stable:
            return "→"
        case .declining:
            return "⬇️"
        case .decliningFast:
            return "⬇️⬇️"
        }
    }
}
